src/vm/fieldmarshaler.cpp

```diff
@@ -1650,9 +1653,13 @@ VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
 
         if (fParentHasLayout)
         {
-            BYTE alignmentRequirement;
-            
-            alignmentRequirement = min(packingSize, pParentLayoutInfo->GetLargestAlignmentRequirementOfAllMembers());
+            BYTE alignmentRequirement = pParentLayoutInfo->GetLargestAlignmentRequirementOfAllMembers();
+
+            // Only take the smaller of the two if we don't contain a SIMD hardware intrinsic type
```

Comment:
This logic does not work for ARM64 Vector64

Reply:
> The comment should definitely be more explicit.

👍 

> I think the logic is correct for ARM32

It look correct for Vector64, but incorrect for Vector128.  The `language alignment modifier` will be incorrectly ignored.

> ARM64 ABI looks to be non-explicit about the handling of these types when a language provided alignment or packing differs from the default.

Agreed.

> The member alignment of an element of a composite type is the alignment of that member after the application of any language alignment modifiers to that member

Since it is reasonable to assume that `language alignment modifiers` should function normally, I would disable these blocks for ARM64 and ARM32.

Either way all Short Vector types need to be treated the same.

My recommendation would be to place all these conditional `16` blocks in `#ifdef _TARGET_XARCH_`

Reply:
> It looks correct for Vector64, but incorrect for Vector128.
`Vector128` on ARM32 is actually 8-byte aligned, so it will be respected here.

Reply:
@sdmaclea, after reading through the ARM32 and ARM64 standards more in depth. It looks like ARM32 is meant to be impacted by the alignment the user specifies. However, ARM64 is not.

Specifically, for ARM32 (they are referred to as `Containerized Vector`):
> The content of a containerized vector is opaque to most of the procedure call standard: the only defined aspect of its layout is the mapping between the memory format (the way a fundamental type is stored in memory) and different classes of register at a procedure call interface. If a language binding defines data types that map directly onto the containerized vectors it will define how this mapping is performed.

For ARM64:
> A short vector is a machine type that is composed of repeated instances of one fundamental integral or floatingpoint type. It may be 8 or 16 bytes in total size. A short vector has a base type that is the fundamental integral or floating-point type from which it is composed, but its alignment is always the same as its total size. The number of elements in the short vector is always such that the type is fully packed.
> ...
> A language binding may define extended types that map directly onto short vectors. Short vectors are not otherwise created spontaneously (for example because a user has declared an aggregate consisting of eight consecutive byte-sized objects).

Additionally (for ARM64) later in the 'Assignment of arguments to registers and stack' it has the following:
> C.4 If the argument is an HFA, an HVA, a Quad-precision Floating-point or Short Vector Type then the NSAA is rounded up to the larger of 8 or the Natural Alignment of the argument’s type.

Reply:
@tannergooding 

This looks more relevant

> The AAPCS64 does not allow exported interfaces to contain packed structures or bit-fields

So looks like align should never be allowed to affect the layout used in the ABI.

Reply:
Yes. Although, I'm not sure if there is a clear distinction between exported/non-exported interfaces in managed code (given reflection/etc).

The same comment also mentions that GCC doesn't respect this and differs from the expected behavior.

Reply:
> GCC doesn't respect this and differs from the expected behavior.

I think this means GCC doesn't enforce the behavior.   It is the programmers responsibility to follow the rules otherwise interop is compromised.

So I think we could safely use GCC or AAPCS64 behavior here.  It sounds like a don't care.

Reply:
There is the case of correctness and the case of interop with native compilers (such as GCC).

I would say, the best route may be correctness for default (packing = 0), but support interop with explicit.

However, that may have back-compat concerns.

It also still isn't clear on the handling of `Vector` types (which, for ARM64 specifies: ` its alignment is always the same as its total size`, but which GCC allows you to override).

Reply:
> There is the case of correctness and the case of interop with native compilers (such as GCC).
>
> I would say, the best route may be correctness for default (packing = 0), but support interop with explicit.

😕 ❓ 

My read is that GCC & MSVC default structure alignments will be correct for AAPCS interop.  

Therefore `Correctness == interop`.

The GCC issue is only around the non compliant use of packed/unaligned structures in the ABI.  AAPCS does not support them, but notes that they can be created by GCC.  I read these as essentially non-standard uses, outside the scope of  AAPCS.

CoreCLR could match GCC or MSVC both are fully capable of generating compliant AAPCS calls.  
I suspect matching GCC would be simpler.

> It also still isn't clear on the handling of Vector types (which, for ARM64 specifies: its alignment is always the same as its total size, but which GCC allows you to override).

This is the same issue as above.  CoreCLR could allow or prevent override.  
+ Programmer shouldn't but can... (GCC)
+ OR Programmer shouldn't therefore can't (MSVC)

I prefer the gcc approach.  It can allow interoperability with hand coded assembly, marshalling, cross platform proxies....

Reply:
I was indicating that, for the ARM ABI, always aligning the types is correct.

However, there is the case of Native interop for types which aren't ABI compliant, but which certain compilers (such as GCC) allow you to create (which is also important).

> CoreCLR could match GCC or MSVC both are fully capable of generating compliant AAPCS calls.

MSVC actually enforces alignments to be respected, even if a containing struct attempts to override it. While GCC/Clang will allow an alignment on one struct to be overridden by the alignment of another containing struct.

That is, for the following code (MSVC, or the equivalent in GCC/Clang):
```
struct align(16) MyStruct16 { int64_t _0; int64_t _1; }
#include <pshpack1.h>
struct align(1) MyStruct1 { int8_t; _0; MyStruct16 _1; }
#include <poppack.h>
```

On MSVC:
* MyStruct1 will be size 32
* MyStruct1._0 will be offset 0
* MyStruct1._1 will be offset 16

On GCC/Clang
* MyStruct1 will be size 17
* MyStruct1._0 will be offset 0
* MyStruct1._1 will be offset 1

Reply:
Reading your last comment again, it looks like we are saying the same thing 😄 

The default should just work with the ABI, but the user should be allowed to override that for better interop with native code.